kart/sqlalchemy/adapter/sqlserver.py

```diff
@@ -219,38 +219,72 @@ def all_v2_meta_items_including_empty(
             sess.execute(
                 f"""
-                SELECT TOP 1 :column_name AS column_name, {cls.quote(g)}.STSrid AS srid, SRS.*
+                SELECT TOP 1 :column_name AS column_name, {cls.quote(g)}.STSrid AS column_srid, {', '.join(srs_table_joins.keys())}
                 FROM {table_identifier}
-                LEFT OUTER JOIN sys.spatial_reference_systems SRS
-                ON SRS.spatial_reference_id = {cls.quote(g)}.STSrid
+                {chr(10).join(f"LEFT OUTER JOIN {table} ON {column} = {cls.quote(g)}.STSrid" for table, column in srs_table_joins.values())}
                 WHERE {cls.quote(g)} IS NOT NULL;
                 """,
-                {"column_name": g},
-            ).fetchone()
-            for g in geom_cols
-        ]
-        ms_spatial_ref_sys = list(filter(None, ms_spatial_ref_sys))  # Remove nulls.
+                {"column_name": g}
+            ).fetchone() for g in geom_cols))
 
         schema = KartAdapter_SqlServer.sqlserver_to_v2_schema(
-            ms_table_info, ms_spatial_ref_sys, id_salt
+            ms_table_info, ms_crs_info, id_salt
         )
         yield "schema.json", schema
 
-        for crs_info in ms_spatial_ref_sys:
-            auth_name = crs_info["authority_name"]
-            auth_code = crs_info["authorized_spatial_reference_id"]
-            if not auth_name and not auth_code:
-                auth_name, auth_code = "CUSTOM", crs_info["srid"]
-            wkt = crs_info["well_known_text"] or ""
-            yield (
-                f"crs/{auth_name}:{auth_code}.wkt",
-                crs_util.normalise_wkt(
-                    crs_util.ensure_authority_specified(wkt, auth_name, auth_code)
-                ),
-            )
+        for column_name, crs_name, wkt in ms_crs_info:
+            if wkt:
+                yield f"crs/{crs_name}.wkt", wkt
+
+    @classmethod
+    def _parse_crs_rows(cls, crs_rows):
+        # Given an iterable of db rows that look like `column_name | column_srid | authority_name | authorized_spatial_reference_id | well_known_text`
+        # (plus some other uninteresting columns) where that information is sourced from sys.spatial_reference_systems,
+        # and possibly includes some extra columns `auth_name | auth_srid | srtext` (plus some other uninteresting columns)
+        # where the information is sourced from the OGR SRS table [schema].spatial_ref_sys...
+        # yields a 3-tuple `(column_name, crs_name, normalized_crs_wkt)` for each row.
+        for crs_row in crs_rows:
+            if not crs_row:
+                continue
+            column_name = crs_row["column_name"]
+            column_srid = crs_row["column_srid"]
+            if not column_srid:
+                continue
+
+            ms_cols = ["authority_name", "authorized_spatial_reference_id", "well_known_text"]
+            ogr_cols = ["auth_name", "auth_srid", "srtext"]
+            for cols in (ms_cols, ogr_cols):
```

Comment:
why not select particular columns rather than `*` and use aliases to normalise the schema of the returned rows? Then this gets less crazy

Reply:
Done.